state_test.go

```diff
@@ -1155,3 +965,3674 @@ func TestMaxSendingAttempts(t *testing.T) {
+			"ArtifactReboot_Leave_00",
+			"ArtifactReboot_Error_00",
+			"ArtifactRollback_Enter_00",
+			"ArtifactRollback",
+			"ArtifactRollback_Leave_00",
+			"ArtifactRollbackReboot_Enter_00",
+			"ArtifactRollbackReboot",
+			"ArtifactVerifyRollbackReboot",
+			"ArtifactRollbackReboot_Leave_00",
+			"ArtifactFailure_Enter_00",
+			"ArtifactFailure",
+			"ArtifactFailure_Leave_00",
+			"Cleanup",
+		},
+		reportsLog: []string{
+			"downloading",
+			"installing",
+			"rebooting",
+			"failure",
+		},
+		testModuleAttr: testModuleAttr{
+			errorStates: []string{"ArtifactReboot_Leave_00"},
+		},
+		installOutcome: successfulRollback,
+	},
+
+	stateTransitionsWithUpdateModulesTestCase{
+		caseName: "Killed in ArtifactReboot_Leave_00",
+		stateChain: []State{
+			&UpdateFetchState{},
+			&UpdateStoreState{},
+			&UpdateAfterStoreState{},
+			&UpdateInstallState{},
+			&UpdateRebootState{},
+			&UpdateVerifyRebootState{},
+			&UpdateAfterRebootState{},
+			&UpdateCommitState{},
+			&UpdateRollbackState{},
+			&UpdateRollbackRebootState{},
+			&UpdateVerifyRollbackRebootState{},
+			&UpdateAfterRollbackRebootState{},
+			&UpdateErrorState{},
+			&UpdateCleanupState{},
+			&UpdateStatusReportState{},
+			&IdleState{},
+		},
+		artifactStateChain: []string{
+			"Download_Enter_00",
+			"Download",
+			"SupportsRollback",
+			"Download_Leave_00",
+			"ArtifactInstall_Enter_00",
+			"ArtifactInstall",
+			"NeedsArtifactReboot",
+			"ArtifactInstall_Leave_00",
+			"ArtifactReboot_Enter_00",
+			"ArtifactReboot",
+			"ArtifactVerifyReboot",
+			"ArtifactReboot_Leave_00",
+			"ArtifactRollback_Enter_00",
+			"ArtifactRollback",
+			"ArtifactRollback_Leave_00",
+			"ArtifactRollbackReboot_Enter_00",
+			"ArtifactRollbackReboot",
+			"ArtifactVerifyRollbackReboot",
+			"ArtifactRollbackReboot_Leave_00",
+			"ArtifactFailure_Enter_00",
+			"ArtifactFailure",
+			"ArtifactFailure_Leave_00",
+			"Cleanup",
+		},
+		reportsLog: []string{
+			"downloading",
+			"installing",
+			"rebooting",
+			"failure",
+		},
+		testModuleAttr: testModuleAttr{
+			spontRebootStates: []string{"ArtifactReboot_Leave_00"},
+		},
+		installOutcome: successfulRollback,
+	},
+
+	stateTransitionsWithUpdateModulesTestCase{
+		caseName: "Error in ArtifactRollback_Leave_00",
+		stateChain: []State{
+			&UpdateFetchState{},
+			&UpdateStoreState{},
+			&UpdateAfterStoreState{},
+			&UpdateInstallState{},
+			&UpdateRebootState{},
+			&UpdateVerifyRebootState{},
+			&UpdateRollbackState{},
+			&UpdateRollbackRebootState{},
+			&UpdateVerifyRollbackRebootState{},
+			&UpdateAfterRollbackRebootState{},
+			&UpdateErrorState{},
+			&UpdateCleanupState{},
+			&UpdateStatusReportState{},
+			&IdleState{},
+		},
+		artifactStateChain: []string{
+			"Download_Enter_00",
+			"Download",
+			"SupportsRollback",
+			"Download_Leave_00",
+			"ArtifactInstall_Enter_00",
+			"ArtifactInstall",
+			"NeedsArtifactReboot",
+			"ArtifactInstall_Leave_00",
+			"ArtifactReboot_Enter_00",
+			"ArtifactReboot",

```

Comment:
I tried, and I got failures. I'm not really sure why, but I suspect it may be the artifact writing which uses the same files or something. This test appears to be mostly I/O bound anyway, due to the database, so adding more cores gave only negligible improvements for me.